components/relay/src/lib.rs

```diff
@@ -0,0 +1,362 @@
+    pub generated_for: String,
+    pub block_list_emails: bool,
+    pub used_on: Option<String>,
+    pub id: i64,
+    pub address: String,
+    pub domain: i64,
+    pub full_address: String,
+    pub created_at: String, // Use String for timestamps for now (or chrono types later)
+    pub last_modified_at: String,
+    pub last_used_at: Option<String>,
+    pub num_forwarded: i64,
+    pub num_blocked: i64,
+    pub num_level_one_trackers_blocked: i64,
+    pub num_replied: i64,
+    pub num_spam: i64,
+}
+
+#[derive(Debug, Serialize)]
+struct CreateAddressPayload<'a> {
+    enabled: bool,
+    description: &'a str,
+    generated_for: &'a str,
+    used_on: &'a str,
+}
+
+impl RelayClient {
+    pub fn new(server_url: String, auth_token: Option<String>) -> Self {
+        Self {
+            server_url,
+            auth_token,
+        }
+    }
+
+    fn build_url(&self, path: &str) -> Result<Url> {
+        Ok(Url::parse(&format!("{}{}", self.server_url, path))?)
+    }
+
+    fn prepare_request(&self, method: Method, url: Url) -> Result<Request> {
+        log::trace!("making {} request to: {}", method.as_str(), url);
+        let mut request = Request::new(method, url);
+        if let Some(ref token) = self.auth_token {
+            request = request.header(header_names::AUTHORIZATION, &format!("Bearer {}", token))?;
+        }
+        Ok(request)
+    }
+
+    pub fn fetch_addresses(&self) -> Result<Vec<RelayAddress>> {
+        let url = self.build_url("/api/v1/relayaddresses/")?;
+        let request = self.prepare_request(Method::Get, url)?;
+
+        let response = request.send()?;
+        log::trace!("response text: {}", response.text());
+
+        let addresses: Vec<RelayAddress> = response.json()?;
+        Ok(addresses)
+    }
+
+    pub fn accept_terms(&self) -> Result<()> {
+        let url = self.build_url("/api/v1/terms-accepted-user/")?;
+        let request = self.prepare_request(Method::Post, url)?;
+
+        let response = request.send()?;
+        log::trace!("response text: {}", response.text());
+        response.require_success()?;
+        Ok(())
+    }
+
+    pub fn create_address(
+        &self,
+        description: &str,
+        generated_for: &str,
+        used_on: &str,
+    ) -> Result<RelayAddress> {
+        let url = self.build_url("/api/v1/relayaddresses/")?;
+
+        let payload = CreateAddressPayload {
+            enabled: true,
+            description,
+            generated_for,
+            used_on,
+        };
+
+        let mut request = self.prepare_request(Method::Post, url)?;
+        request = request.json(&payload);
+
+        let response = request.send()?;
+        log::trace!("response text: {}", response.text());
+
+        let address: RelayAddress = response.json()?;
+        Ok(address)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use mockito::mock;
+
+    fn base_addresses_json(extra_fields: &str) -> String {
+        format!(
+            r#"
```

Comment:
fyi, the `json!()` macro can be convenient for directly building json in tests etc, which can avoid the need to manually serialize and allows expressions for values.